File: 3757.number-of-effective-subsequences.cpp
#
# @lc app=leetcode id=3757 lang=cpp
#
# [3757] Number of Effective Subsequences
#

# @lc code=start
class Solution {
public:
    int countEffective(vector<int>& nums) {
        const int MOD = 1e9 + 7;
        int fullOR = 0;
        for (int num : nums) {
            fullOR |= num;
        }
        
        // Store frequency of each bit in nums
        vector<int> bitFrequency(32, 0);
        for (int num : nums) {
            for (int i = 0; i < 32; ++i) {
                if (num & (1 << i)) {
                    bitFrequency[i]++;
                }
            }
        }

        long long effectiveCount = 0;

        // Calculate effective subsequences by considering contributions of each bit position
        for (int i = 0; i < 32; ++i) {
            if ((fullOR & (1 << i)) && bitFrequency[i] > 0) { // Check if this bit contributes to fullOR
                // Calculate number of ways we can choose subsets from all numbers having this particular bit set
                long long totalSubsets = (1LL << bitFrequency[i]) % MOD;
                // Subtract subsets that do not contribute to reducing the OR, which are when no element with this bit is present 
                long long ineffectiveSubsets = (totalSubsets - 1 + MOD) % MOD; 
                effectiveCount = (effectiveCount + ineffectiveSubsets) % MOD;
            }
        }

        return static_cast<int>(effectiveCount);
    } 
}; 
# @lc code=end

File: 3757.number-of-effective-subsequences.java
#
# @lc app=leetcode id=3757 lang=java
#
# [3757] Number of Effective Subsequences
#

# @lc code=start
import java.util.*;
class Solution {
    public int countEffective(int[] nums) {
        final int MOD = 1000000007;
        int n = nums.length;
        int totalOR = 0;
        for (int num : nums) totalOR |= num;
        List<Integer> bits = new ArrayList<>();
        for (int b = 0; b < 21; ++b) if (((totalOR >> b) & 1) != 0) bits.add(b);
        int m = bits.size();
        List<List<Integer>> bitIndices = new ArrayList<>();
        for (int i = 0; i < m; ++i) bitIndices.add(new ArrayList<>());
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (((nums[i] >> bits.get(j)) & 1) != 0) bitIndices.get(j).add(i);
            }
        }
        int res = 0;
        for (int mask = 1; mask < (1 << m); ++mask) {
            Set<Integer> indices = null;
            for (int j = 0; j < m; ++j) {
                if (((mask >> j) & 1) != 0) {
                    if (indices == null) indices = new HashSet<>(bitIndices.get(j));
                    else indices.retainAll(bitIndices.get(j));
                }
            }
            if (indices == null || indices.isEmpty()) continue;
            int cnt = indices.size();
            long ways = pow(2, cnt, MOD) - 1;
            if (ways < 0) ways += MOD;
            int bitsOn = Integer.bitCount(mask);
            if ((bitsOn & 1) == 1) res = (int)((res + ways) % MOD);
            else res = (int)((res - ways + MOD) % MOD);
        }
        return res;
    }
    private long pow(long base, int exp, int mod) {
        long ans = 1;
        while (exp > 0) {
            if ((exp & 1) == 1) ans = ans * base % mod;
            base = base * base % mod;
            exp >>= 1;
        }
        return ans;
    }
}
# @lc code=end